modelservice/worker.py

```diff
@@ -72,12 +78,118 @@ def register(self):
+        normalize_inputs: bool = hyper_parameters.normalize_inputs
+        train_split: float = hyper_parameters.train_split
+        test_split: float = 1.0 - train_split
+
+        if hyper_parameters.optimizer_type == OptimizerType.ADAM:
+            optimizer = tf.keras.optimizers.Adam(learning_rate)
+        else:
+            optimizer = tf.keras.optimizers.SGD(learning_rate)
+
+        if hyper_parameters.loss_type == LossType.MEAN_SQUARED_ERROR:
+            loss = "mean_squared_error"
+        elif hyper_parameters.loss_type == LossType.MEAN_SQUARED_ERROR:
+            loss = "mean_squared_error"
+        else:
+            loss = "mean_absolute_error"
+
+        evaluation_metrics: list = []  # list(EvaluationMetric)
+
+        # Make models dir for job id
+        models_dir: str = f"{self.data_dir}/{request.id}"
+        os.mkdir(models_dir)
+
+        count = 1
+        gis_join_timer: Timer = Timer()
+        for gis_join in request.gis_joins:
+
+            gis_join_timer.start()
+            info(f"Loading data for GISJOIN {gis_join} ({count}/{len(request.gis_joins)})...")
+
+            # Load data
+            csv_path: str = f"{self.data_dir}/{gis_join}.csv"
+            all_df: pd.DataFrame = pd.read_csv(csv_path, header=0).drop("GISJOIN", 1)
+            len_df: int = len(all_df.index)
+            info(f"Loaded data for GISJOIN {gis_join}: {len_df} records")
+            if normalize_inputs:
+                scaled = MinMaxScaler(feature_range=(0, 1)).fit_transform(all_df)
+                all_df = pd.DataFrame(scaled, columns=all_df.columns)
+
+            features = all_df[feature_fields]
+            labels = all_df[label_field]
+
+            # Create Sequential model
+            model = tf.keras.Sequential()
+
+            # Add input layer
+            model.add(tf.keras.Input(shape=(len(request.feature_fields))))
+
+            # Add hidden layers
+            for hidden_layer in hyper_parameters.hidden_layers:
+                name: str = hidden_layer.name
+                units: int = hidden_layer.units
+                activation = "relu"
+                model.add(tf.keras.layers.Dense(units=units, activation=activation, name=name))
+
+            # Add output layer
+            output_layer: OutputLayer = hyper_parameters.output_layer
+            name: str = output_layer.name
+            activation = "relu"
+            model.add(tf.keras.layers.Dense(units=1, activation=activation, name=name))
+
+            # Compile the model and print its summary
+            model.compile(loss=loss, optimizer=optimizer)
+            model.summary()
+
+            # Fit the model to the data
+            history = model.fit(features, labels, epochs=epochs, validation_split=test_split)
+            hist = pd.DataFrame(history.history)
+            hist["epoch"] = history.epoch
+            info(hist)
+
+            last_row = hist.loc[hist["epoch"] == epochs - 1].values[0]
```

Comment:
I'm not super familiar with pandas. How does passing a single boolean value to `.loc` work? I thought you had to pass an array with length equal to the number of rows when using boolean values.

Reply:
This works as a filter, but you're right we should be using this: https://stackoverflow.com/questions/70811348/python-get-last-row-from-pandas-data-frame